qubesadmin/tools/qvm_template.py

```diff
@@ -1574,6 +1580,56 @@ def repolist(args: argparse.Namespace, app: qubesadmin.app.QubesBase) -> None:
         qubesadmin.tools.print_table(table)
 
 
+def migrate_from_rpmdb(app):
+    """Migrate templates stored in rpmdb, into 'features' set on the VM itself.
+    """
+
+    if os.getuid() != 0:
+        parser.error('This command needs to be run as root')
+    rpm_ts = rpm.TransactionSet()
+    pkgs_to_remove = []
+    for pkg in rpm_ts.dbMatch():
+        if not pkg[rpm.RPMTAG_NAME].startswith('qubes-template-'):
+            continue
+        try:
+            vm = app.domains[pkg[rpm.RPMTAG_NAME][len('qubes-template-'):]]
+        except KeyError:
+            # no longer present, remove from rpmdb
+            pkgs_to_remove.append(pkg)
+            continue
+        if is_managed_template(vm):
+            # already migrated, cleanup
+            pkgs_to_remove.append(pkg)
+            continue
+        try:
+            vm.features['template-name'] = vm.name
+            vm.features['template-epoch'] = pkg[rpm.RPMTAG_EPOCHNUM]
+            vm.features['template-version'] = pkg[rpm.RPMTAG_VERSION]
+            vm.features['template-release'] = pkg[rpm.RPMTAG_RELEASE]
+            vm.features['template-reponame'] = '@commandline'
+            vm.features['template-buildtime'] = \
+                datetime.datetime.fromtimestamp(
+                    pkg[rpm.RPMTAG_BUILDTIME], tz=datetime.timezone.utc).\
+                strftime(DATE_FMT)
+            vm.features['template-installtime'] = \
+                datetime.datetime.fromtimestamp(
+                    pkg[rpm.RPMTAG_INSTALLTIME], tz=datetime.timezone.utc).\
+                strftime(DATE_FMT)
+            vm.features['template-license'] = pkg[rpm.RPMTAG_LICENSE]
+            vm.features['template-url'] = pkg[rpm.RPMTAG_URL]
+            vm.features['template-summary'] = pkg[rpm.RPMTAG_SUMMARY]
+            vm.features['template-description'] = \
+                pkg[rpm.RPMTAG_DESCRIPTION].replace('\n', '|')
+            vm.installed_by_rpm = False
+        except Exception as e:  # pylint: disable=broad-except
+            print('Failed to set template {} metadata: {}'.format(vm.name, e))
+            continue
+        pkgs_to_remove.append(pkg)
+    subprocess.check_call(
+        ['rpm', '-e', '--justdb'] +
```

Comment:
```suggestion
        ['rpm', '-e', '--justdb', '--'] +
```

in case the package name starts with `-` (it shouldn’t, but better to be safe)

Reply:
That's impossible - the loop above explicitly collects only packages with `qubes-template-` prefix.

Reply:
That wasn’t obvious from looking at the code.  Would it be possible to collect a list of strings instead of a list of package objects, and add an assertion here?